### ISSUE TITLE
Cardinality for mutually-recursive types

### ISSUE THREAD

**Issue by Bodigrim** [identicalsnowflake/cantor-pairing#6]
Current `HEAD` fails to detect cardinality of mutually-recursive types as `Countable`. Here is an example:

```haskell
{-# LANGUAGE DeriveGeneric, TypeApplications #-}

import Cantor
import GHC.Generics

data Foo = FooNil | Foo Bool Bar deriving (Generic)
data Bar = BarNil | Bar Bool Foo deriving (Generic)

instance Cantor Foo
instance Cantor Bar

main = print $ cardinality @Foo
```

```
$ ghc MutualRec.hs && ./MutualRec
MutualRec: <<loop>>
```

I believe that this can be improved by keepind track of all encountered types (instead of only the last one) in the first argument of `GCantor` class, as a type-level list. Does it look worse pursuing?

**Comment by identicalsnowflake** [comment]
Yeah currently it only does simple inductive types correctly. I would be happy to support more stuff, like mutually recursive types or GADTs, I'm just not sure if it's worth the effort.

My experience in trying to use instance resolution to do fancy things is mostly bad, in that it's super obvious that this sort of thing should be possible, but when I actually sit down to try to implement it, it ends up being an enormous pain in the ass. It feels like programming in a crappy version of the untyped lambda calculus.

Anyway, if you want to take a shot at it, I'll be happy to merge the PR :wink:

**Comment by identicalsnowflake** [comment]
By the way, you can manually specify the cardinality as `Countable` and still get the everything else for free:

```haskell
data Foo = FooNil | Foo Bool Bar deriving (Generic,Show)
data Bar = BarNil | Bar Bool Foo deriving (Generic,Show)

instance Cantor Foo where
  cardinality = Countable
instance Cantor Bar
```

I should probably add this to the docs somewhere.

EDIT: done

**Comment by Bodigrim** [comment]
I gave it a try and it seems not worth it. `GCantor` instances even in their current form look pretty scary :) And tracking more stuff on the type level makes type checking extremely fragile. It is probably a more robust solution to define `cardinality` manually, as you suggested, rather than depend on such thin matters.

Feel free to reopen, if you decide otherwise.

**Comment by identicalsnowflake** [comment]
`GCantor` in general is a bit hairy, but I think the only part relevant to this issue is `gcardinality`, which isn't bad.

It may possible to just factor the cardinality out into its own `GCardinality` class, which lets that set of instances exist in its own self-contained universe where it can be reasoned about in isolation without the distractions of the encoding mechanisms of the library.
